Move all Blockscout config parameters to Helm

carterqw2 opened the pull request:
### Description

Changes:
- All config parameters are moved from `.env` files to helm values files.
- Secrets are moved from `.env` files to Secret Manager (managed manually currently).
The final goal is to be able to use helm in CI/CD pipelines for Blockscout deployments.

### Other changes

Updates the `deploy initial` command to create required secrets in Secret Manager.
Updates the `deploy destroy` command to delete secrets created previously.

### Tested

Every commit in this PR is tested against every deployment.
Note to reviewers: please decrypt the secrets and make sure there are no `BLOCKSCOUT_` secrets in decrypted `.env` files.

### Related issues

- Fixes https://github.com/celo-org/data-services/issues/97
- Fixes https://github.com/celo-org/data-services/issues/249
- Fixes https://github.com/celo-org/data-services/issues/72

### Backwards compatibility

`celotool` still can be used for deployments with the same syntax. The config file of k8s DB secrets was removed, so the previous version of `celotool` won't be able to read them.

### Documentation

[Secret Management in Blockscout How-to](https://www.notion.so/clabsco/Secret-Management-in-Blockscout-3d2e6a1081f54b20b3ff6017150f9fc3)

Commit:
Removed BLOCKSCOUT_DOCKER_IMAGE_REPOSITORY

Commit:
Reordered metadata-crawler's config

Commit:
Removed BLOCKSCOUT_SUBNETWORK_NAME

Commit:
Reorganized usage of NETWORK_ID

Commit:
Removed metadata crawler variables

Commit:
Removed archive nodes variables

Commit:
Removed ability to drop the db directly via celotool cli

Commit:
Removed domain name param

Commit:
Updated change cause param

Commit:
Removed db connection name parameter

Commit:
Camelcased all parameter names

Commit:
Refactored helm templates and added secret-init

Commit:
Removed segment key secret

Commit:
Removed discord webhook url

Commit:
Removed db secrets

Commit:
Fixed deprecation warning

Commit:
Removed recaptcha secrets

Commit:
Annotated the service account

Commit:
Updated deploy initial command

Commit:
Fixed linting

Commit:
Cleanup of mnemonic files

Commit:
Updated destroy command

rkachowski commented on packages/celotool/src/lib/blockscout.ts:
```diff
@@ -165,7 +71,7 @@ export async function createDefaultIngressIfNotExists(celoEnv: string, ingressNa
     console.info(`Creating ingress ${celoEnv}-blockscout-web-ingress`)
     const ingressFilePath = `/tmp/${celoEnv}-blockscout-web-ingress.yaml`
     const ingressResource = `
```
Is there no way to remove the ingress resource js template and use a regular helm template?

carterqw2 replied:
Looks like it's done intentionally so it doesn't override the existing ingress that points to one of the two existing instances. Since we had an intention to unify deployments anyway, this would be removed anyway.

lvpeschke commented on packages/helm-charts/blockscout/values-rc1staging.yaml:
```diff
@@ -48,7 +48,7 @@ blockscout:
     recaptchaSecretName: rc1staging-blockscout-recaptcha
   metadata_crawler:
-    schedule: "0 */2 * * *"
+    enabled: true
```
Why do we want this for staging? (same as for blockscoutstagingrc1)

carterqw2 replied:
I just reflected what we had on different envs but I agree that we can remove that on stagings. Done.

lvpeschke commented on packages/helm-charts/blockscout/values-blockscoutstagingrc1.yaml:
```diff
@@ -49,6 +49,8 @@ blockscout:
         cpu: 500m
     recaptchaSecretName: blockscoutstagingrc1-blockscout-recaptcha
+  metadata_crawler:
+    enabled: true
```
Why do we want this for staging?

lvpeschke commented on packages/helm-charts/blockscout/values.yaml:
```diff
@@ -166,7 +166,8 @@ blockscout:
   metadata_crawler:
-    schedule: "*/30 * * * *"
+    enabled: false
+    schedule: "0 */2 * * *"
```
Could you add a human friendly comment to quickly parse what this means? Here and elsewhere.
(i.e. "Every 2h")

carterqw2 replied:
Done!

lvpeschke commented on packages/helm-charts/blockscout/values.yaml:
```diff
@@ -171,6 +171,9 @@ blockscout:
       tag: metadata-crawler-77a392216d4927e85ce4b683508fc0539aa92a34
+  archiveNodes:
+    jsonrpc_http_url: http://tx-nodes-private:8545
+    jsonrpc_ws_url: ws://tx-nodes-private:8545
```
💯  to this style of config!

rkachowski replied:
this is an intermediate change, right? the final var is in camel case - `jsonrpcHttpUrl` (???)

lvpeschke replied:
(FWIW: my comment was about the fact that it was now possible to understand which archive node urls are used by reading the helm config, instead of the helm config + celotool + env variables; +1 to camel case which is the final form, I did a commit per commit review)

lvpeschke commented on packages/helm-charts/blockscout/templates/_helpers.tpl:
```diff
@@ -44,15 +44,44 @@ the `volumes` section.
+  args:
+  - job
+  - {{ .Release.Name }}-migration
+{{- end -}}
```
Is this required given the Helm hooks logic?

eruizgar91 replied:
It's not need it anymore https://github.com/celo-org/celo-monorepo/pull/9156/files @lvpeschke

lvpeschke replied:
My bad, I was looking at both at the same time!

lvpeschke commented on packages/celotool/src/lib/helm_deploy.ts:
```diff
@@ -152,6 +163,28 @@ export async function createCloudSQLInstance(celoEnv: string, instanceName: stri
+    await execCmd(`gcloud secrets describe ${secretId}`)
+
+    console.log(`Secret ${secretId} already exists, skipping creation...`)
+  } catch (error) {
```
Not super clean, but I'd add a check for `helmdryrun` here to prevent creating a new secret when having that flag set.

carterqw2 replied:
The check is performed in the [command handler](https://github.com/celo-org/celo-monorepo/pull/9140/commits/91f45183a283188620f4384213cb5ec1239cbd5f#diff-5d700e66c8eedb29450acb4427a80e0d6518c5a99efb239d90a0c5b482cf60c0R33). This file is a library that performs final actions.

lvpeschke commented on packages/celotool/src/lib/helm_deploy.ts:
```diff
@@ -152,6 +163,28 @@ export async function createCloudSQLInstance(celoEnv: string, instanceName: stri
+    await execCmd(
+      `echo -n "${secretValue}" | gcloud secrets versions add ${secretId} --data-file=-`
+    )
+  }
```
Is there an advantage in splitting this across 2 invocations instead of 1?

carterqw2 replied:
Good point, I misread the doc. Thank you, updated.

lvpeschke commented on packages/celotool/src/lib/helm_deploy.ts:
```diff
@@ -185,6 +185,14 @@ export async function createSecretInSecretManagerIfNotExists(
   }
 }
 
+export async function deleteSecretFromSecretManager(secretId: string) {
```
I would add a prompt to confirm deletion to lower the chances for accidents.

carterqw2 replied:
There is an existing [check](https://github.com/celo-org/celo-monorepo/blob/master/packages/celotool/src/lib/env-utils.ts#L351) that asks for confirmation for any change on production environments. Do you think it's not enough?

lvpeschke replied:
Missed updating this one: I would argue it's not enough because that is at the very beginning of the operation. Adding a prompt here doesn't save everything, as deleting the DB and other things could also be qualified as "accidents", but I expect secrets to be more locked down than other resources for security. Not a blocker for deployment though, just a strong preference.

lvpeschke reviewed (commented):
Super cool, I appreciate how this makes the config more readable in so many places!
Thanks for providing easy to follow commits for this rather huge change!
I've left a few comments and questions, but overall this looks very good.

rkachowski commented on packages/helm-charts/blockscout/templates/_helpers.tpl:
```diff
@@ -12,7 +12,7 @@ heritage: {{ .Release.Service }}
 */ -}}
 {{- define "celo.blockscout.annotations" -}}
-kubernetes.io/change-cause: Deployed {{ .Values.blockscout.image.tag }} by {{ .Values.blockscout.deployment.account }} on {{ .Values.blockscout.deployment.timestamp }}
+kubernetes.io/change-cause: {{ .Values.changeCause }}
```
I like moving the string interpolation out of the template and into the deployment code.

rkachowski commented on packages/helm-charts/blockscout/templates/blockscout-api.ingress.yaml:
```diff
@@ -19,13 +19,16 @@ metadata:
     secretName: {{ .Release.Name }}-api-tls
   rules:
-  - host: {{ .Release.Name }}-api.{{ .Values.domain.name }}.org
+  - host: {{ .Release.Name }}-api.{{ .Values.blockscout.domainName }}
```
`domainName` now includes the tld whilst previously it was just the host?

carterqw2 replied:
Yeah, I don't have a good explanation why we need to hardcode the `.org` part.

jcortejoso replied:
My impressing is that originally the domain could be the same as the like `project_name` but eventually this was broke.

rkachowski commented on packages/helm-charts/blockscout/templates/blockscout-indexer.deployment.yaml:
```diff
@@ -86,12 +89,15 @@ spec:
-{{- $data := dict "Values" .Values "Database" .Values.blockscout.indexer.db }}
-{{ include "celo.blockscout-db-sidecar" $data | indent 6 }}
+{{- $data := dict "Release" .Release "Values" .Values "Database" .Values.blockscout.indexer.db }}
+{{ include "celo.blockscout.env-vars" $data | indent 8 }}
```
this is perhaps outwith the scope of this PR, but I feel we should at some point enforce a convention for helm template var style - older / legacy values seem to have a hyphenated `env-vars` style, whilst our most recent changes are `poolSize` camel case.

carterqw2 replied:
In this case `env-vars` is the name of a defined template. Template filenames [use dashes](https://v2.helm.sh/docs/chart_best_practices/#templates) according to best practices as well as [chart names](https://helm.sh/docs/chart_best_practices/conventions), so I thought defined templates should use them too.

rkachowski commented on packages/helm-charts/blockscout/templates/blockscout-web.deployment.yaml:
```diff
@@ -113,23 +116,19 @@ spec:
-            secretKeyRef:
-              name:  {{ .Values.blockscout.web.recaptchaSecretName | quote }}
-              key: project_id
+          value: {{ .Values.blockscout.secrets.recaptcha.projectId }}
```
Q for my understanding: this was previously a single k8s secret which was essentially a map of 3 values indexed by the `key` to retrieve the necessary credential. Are these now individual secrets within gcp's credential management system?

carterqw2 replied:
Yes, that's right, there are three different secrets in GCP Secret Manger. I thought we could ditch `projectId` and `siteKey` as they are rendered in open text anyway but left for completeness.

rkachowski commented on packages/helm-charts/blockscout/values.yaml:
```diff
@@ -166,21 +186,40 @@ blockscout:
-    image:
-      repository: gcr.io/celo-testnet/celo-monorepo
-      tag: metadata-crawler-77a392216d4927e85ce4b683508fc0539aa92a34
+  secrets:
```
it's nice to see the required credentials in such a concise format

rkachowski commented:
It looks pretty good to me, I added some comments + questions mostly for my understanding.

edit: whoops, clicked close with comment by mistake

Commit:
Add migration hook (#9156)

Commit:
Fixed PR comments

Commit:
Add grafana hook (#9126)

lvpeschke commented on packages/helm-charts/blockscout/templates/grafana-hook.yaml:
```diff
@@ -0,0 +1,50 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: "{{ .Release.Name }}"
```
Could this be made more specific? eg with a suffix to make it obvious what this is

lvpeschke reviewed (approved):
This has grown into quite a huge change; I might have overlooked something. However, it definitely passes the test of "does this improve the overall situation?", so LGTM.

Commit:
Made the annotation post install hook name more specific

Commit:
Fixed a typo

Commit:
Turned service account into a hook

Commit:
Updated pathType to match original ingress files

Commit:
Removed DATABASE_URL variable

Commit:
Fixed GCP-related values for blockscoutstagingrc1

Commit:
Fixed deprecation warning for CronJob

Commit:
Moved migration job removal logic to helm hook

Commit:
Fixed grafana cloud annotations hook

Made the script config map static so it's always overwritten.
Updated the hook's deletion policy so it get's overwritten on re-deploys
and gets deleted on success.

lvpeschke commented on packages/helm-charts/blockscout/templates/_helpers.tpl:
```diff
@@ -145,8 +145,6 @@ blockscout components.
   value: {{ .Values.blockscout.archiveNodes.jsonrpcWsUrl }}
 - name: PGUSER
   value: {{ .Values.blockscout.secrets.dbUser }}
-- name: DATABASE_URL
```
How come this can be removed?

carterqw2 replied:
We build the connection URL from individual parts instead, since we can't use string interpolation for secrets.

lvpeschke commented on packages/celotool/src/cmds/deploy/upgrade/blockscout.ts:
```diff
@@ -61,14 +60,6 @@ export const handler = async (argv: BlockscoutUpgradeArgv) => {
-        console.error(error)
-      }
     }
   } else {
```
Can you change the wording in this else block or remove it?

lvpeschke commented on packages/helm-charts/blockscout/values-rc1staging-blockscout.yaml:
```diff
@@ -36,31 +38,36 @@ blockscout:
+    jsonrpcHttpUrl: http://rc1staging-archivenodes-rpc:8545
+    jsonrpcWsUrl: ws://rc1staging-archivenodes-rpc:8545
+  metrics:
+    enabled: false
```
This should be true?

Commit:
Refactored volumes

Commit:
Turned service account, role and binding into a hook

Commit:
Fix install values name

Commit:
Update values to new deployment blockscoutstagingrc1

Commit:
Updated data migration job definition

Commit:
Updated data migration job definition part 2

Commit:
Added helm timeout param to wait for hooks to finish

jcortejoso commented on packages/celotool/src/cmds/deploy/initial/blockscout.ts:
```diff
@@ -43,24 +38,25 @@ export const handler = async (argv: InitialArgv) => {
 
-    blockscoutCredentials = await createCloudSQLInstance(argv.celoEnv, instanceName)
+    const [dbUser, dbPassword, _] = await createCloudSQLInstance(argv.celoEnv, instanceName)
+    const secretLabels = ['app=blockscout', `env=${argv.celoEnv}`]
```
Include `team` = `data-services` as label?

jcortejoso commented on packages/celotool/src/cmds/deploy/migrate/blockscout.ts:
```diff
@@ -1,7 +1,6 @@
-import { getInstanceName, getReleaseName, installHelmChart } from 'src/lib/blockscout'
+import { getReleaseName, installHelmChart } from 'src/lib/blockscout'
```
What is the function of `migrate` command? Is it for installing a "new" blockscout release when the secret already exits?

carterqw2 replied:
It's hard to say, we've never used it really.

jcortejoso commented on packages/helm-charts/blockscout/scripts/grafana-hook.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
😍

jcortejoso reviewed (approved):
Nice PR

Commit:
Cleanup

Commit:
Updated cookie env var format

Commit:
Switched web pods to db replica

Commit:
Updated HPA configs

Commit:
Service account updates

Commit:
Added makerdorjo URL for rc13

Commit:
Added team label to the secret

Commit:
Removed unnecessary else statement

Commit:
Merge branch 'master' into carterqw2/blockscout-config-helm

Commit:
Added erlang cookie secret variables

Commit:
Hooks order and garbage collection

Commit:
Updated web db instance

Commit:
Merge branch 'master' into carterqw2/blockscout-config-helm

lvpeschke commented on packages/celotool/src/lib/helm_deploy.ts:
```diff
@@ -1000,7 +1037,7 @@ export async function upgradeGenericHelmChart(
     console.info(`Upgrading helm release ${releaseName}`)
     await helmCommand(
-      `helm upgrade --install ${valuesOverride} ${releaseName} ${chartDir} --namespace ${namespace} ${parameters.join(
+      `helm upgrade --install ${valuesOverride} ${releaseName} ${chartDir} --timeout 120h --namespace ${namespace} ${parameters.join(
```
Any reason for having such a high timeout here? Could any lower value be enough?

carterqw2 replied:
Based on my experience, we've had very long-running migrations in the past and since migrations run in a hook now, we have to be conservative here.

lvpeschke reviewed (approved):
Very nice!

Commit:
Fixed build

Commit:
Removed unused variable
